WIP: Raft tracing

andrewbaptist opened the pull request.

Commit:
raft: add tracing to raft messages

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
raft: add test

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
raft: convert messages to ContextMessages

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
raft: pass context through everywhere

This commit converts Raft to use the standard logging approach of
passing a context.Context to all the methods and using the CRDB logger.
This will allow structure logging through Raft.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131207)
<!-- Reviewable:end -->

Commit:
temp: until raft tracing is sorted

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
TESTING ONLY: Add strict checks for Raft messages

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Commit:
raft: add trace encoding to raft.Entry

Change the encoding of raft to add whether the Entry is tracked. This
uses a previously unused bit within the first byte of the Entry.

TODO: Validate compatibility considerations.

Epic: none

Release note: None

Commit:
Use proposal buf map
